cmd/deploy.go

```diff
@@ -7,62 +7,109 @@ import (
 	"github.com/spf13/cobra"
 
 	"github.com/boson-project/faas"
+	"github.com/boson-project/faas/buildpacks"
 	"github.com/boson-project/faas/docker"
 	"github.com/boson-project/faas/knative"
+	"github.com/boson-project/faas/progress"
 	"github.com/boson-project/faas/prompt"
 )
 
 func init() {
 	root.AddCommand(deployCmd)
 	deployCmd.Flags().BoolP("confirm", "c", false, "Prompt to confirm all configuration options - $FAAS_CONFIRM")
+	deployCmd.Flags().StringP("image", "i", "", "Optional full image name, in form [registry]/[namespace]/[name]:[tag] for example quay.io/myrepo/project.name:latest (overrides --repository) - $FAAS_IMAGE")
 	deployCmd.Flags().StringP("namespace", "n", "", "Override namespace into which the Function is deployed (on supported platforms).  Default is to use currently active underlying platform setting - $FAAS_NAMESPACE")
 	deployCmd.Flags().StringP("path", "p", cwd(), "Path to the function project directory - $FAAS_PATH")
+	deployCmd.Flags().StringP("repository", "r", "", "Repository for built images, ex 'docker.io/myuser' or just 'myuser'.  - $FAAS_REPOSITORY")
 }
 
 var deployCmd = &cobra.Command{
 	Use:   "deploy",
 	Short: "Deploy an existing Function project to a cluster",
 	Long: `Deploy an existing Function project to a cluster
 
-Deploys the Function project in the current directory. A path to the project
-directory may be provided using the --path or -p flag. The image to be deployed
-must have already been created using the "build" command.
+Builds and Deploys the Function project in the current directory. 
```

Comment:
From this description it looks like that `faas build` and `faas deploy` overlap, so that `deploy` is a superset of `build`. This is fine, but then it would be better to include the _build only_ mode as an option to `faas deploy`, like in `faas deploy --build-only`. This would have several benefits:

* Reducing duplication in options that are the same for `deploy` and `build`
* Less commands, smaller UI interface

Drawbacks are:

* Maybe not easy to find out which options of `faas deploy` are relevant for build and which for creating the service.
* Build can be considered as a high-level concern, so a dedicated command might reflect this more prominently

I'm not sure what's the best way to go. From my experience with `serverless.com` integration, I can say that "Faas" platforms don't expose the "build" aspect so prominently as many FaaS platform don't have that stage anyway (think AWS Lambda). So it's just an aspect of the one single, most important operation: How to get my local code running in the cloud ? Aka "deploy". From that angle, I'd recommend to fold "build" as just another aspect of "deploy" and enable this with a dedicated option that influence the flow (its a bit similar to `--dry-run` option that is used by many commands)

On the other hand, if we really want to expose the build as a first level concept, which e.g. is also used *without* deploying (eg for local development with _containers_), then this separate commands make definitely sense.